### ISSUE TITLE
gazctl: shards prune

### ISSUE THREAD

**Issue by jgraettinger** [LiveRamp/gazette#101]
As an operator, I want to provide a shard label selector, and have journal fragments of matched ShardSpec recovery logs that are not required for recovering from any of a shard’s current hints, to be deleted.

(Make gRPC api for fetching current hints?)

**Comment by jskelcy** [comment]
I have started looking into this and I am looking at the [v1 implementation](https://github.com/LiveRamp/gazette/blob/master/cmd/gazctl/cmd/shard_prune_log.go) for inspiration and I am seeking some guidance from people with more experice with gazette v2 than me. It appears that the v1 implementation relies on the [`FileSystem`](https://github.com/LiveRamp/gazette/blob/master/pkg/cloudstore/filesystem.go) interface for:

1.  Walking the list of fragments available.
2. Deleting fragments which contain are no longer relevant. 

While many packages in v1 have been ported to v2 it appears that the `FileSystem` interface, and the entire `cloudstore` pacakge has not. My question is has the `FileSystem` interface (and relevant implementations) not been move because there is an analogous interface I have overlooked, or has it not been moved becaause it has not needed to be moved yet?

As far as walking the list of available fragments it would appear that using the [reader client](https://github.com/LiveRamp/gazette/blob/master/v2/pkg/client/reader.go#L36) with a starting offset of zero may be equivalent to using [`Walk`](https://github.com/LiveRamp/gazette/blob/master/pkg/cloudstore/filesystem.go#L73) function provided by the `FileSystem` interface. The only difference being that after all the of the fragments that have been persisted to cloud storage have been inspected fragments which are still being activly written to will be inspected, however I am not sure that is an issue as we can detect that this fragment has not been flushed and if so not consider it for deletion. The second component of this the prune tool seems to be a bigger issue, where there is not a good analog for deleting fragments which are no longer relevant. 

If I am missing something please let me know, or if this is a good opportunity to port the `cloudstorage` package to v2 we can take a look at that too.

**Comment by jskelcy** [comment]
I spoke with Johnny offline and it sounds like there were significant usability issues with the previous cloud storage interface so it will not be ported. Most of the functionality that was made available via `cloudstorage.FileSystem` is available in the `Fragment` package in the file `stores.go`. 

This diff adds the ability to remove fragments which is missing from the cloud stage abstraction in gazette v2:
https://github.com/LiveRamp/gazette/pull/133

**Comment by jskelcy** [comment]
There is a change to consider here between the functionality of v1 and v2 that needs to be discussed. In v1 users were asked to provide etcd keys from which we would fetch hints. However, I think moving forward :
- It will be safer to not query etcd directly from the cli tool but rather provide an endpoint as part of the consumer grpc interface to return hints.
- Rather than asking users for a etcd key to fetch hints, instead ask them for a selector to resolve consumer shards. 

I think the only meaningful change here is that this tool becomes much more powerful because it can be run once per "topic", where as before it needed to be run once per consumer shard (as I understand it). One difference here is that a run of this tool may now take much longer as running a single command may prune many recovery logs at once. 

If this raises any red flags for anyone please let me know.

**Comment by jskelcy** [comment]
After some discussion, we have learned that this task is blocked by #131. This task will be resumed when that ticket has been resolved.

**Comment by jskelcy** [comment]
This ticket involves inspecting hints from the consumer etcd and is blocked by #146
